presto-main/src/main/java/com/facebook/presto/server/protocol/QueryRateLimiter.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server.protocol;
+
+import com.facebook.presto.execution.QueryManagerConfig;
+import com.facebook.presto.spi.QueryId;
+import com.google.inject.Inject;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class QueryRateLimiter
+{
+    private final Map<QueryId, ConcurrentLinkedQueue<Long>> rateLimitMap = new ConcurrentHashMap<>();
+
+    private final long rateLimitBucketMaxSize;
+    private final long rateLimitDuration;
+    private final boolean rateLimitEnabled;
+
+    @Inject
+    public QueryRateLimiter(QueryManagerConfig queryManagerConfig)
+    {
+        requireNonNull(queryManagerConfig, "queryManagerConfig is null");
+        this.rateLimitBucketMaxSize = queryManagerConfig.getRateLimitBucketMaxSize();
+        this.rateLimitDuration = (long) 1E9; // #todo config
+        this.rateLimitEnabled = true; // #todo config
+        checkArgument(rateLimitBucketMaxSize >= 1, "Rate limit token bucket max size should be at least 1");
+    }
+
+    public boolean shouldRateLimitQuery(QueryId queryId)
+    {
+        //System.out.println("Starting shouldRateLimit");
+        requireNonNull(queryId, "queryId cannot be null");
+        ConcurrentLinkedQueue<Long> tokenBucketPermits = rateLimitMap.get(queryId);
+        if (tokenBucketPermits == null) {
+            tokenBucketPermits = rateLimitMap.computeIfAbsent(queryId, notused -> new ConcurrentLinkedQueue<Long>());
+        }
+        synchronized (tokenBucketPermits) {
+            tokenBucketPermits.add(System.nanoTime());
+            if (tokenBucketPermits.size() <= rateLimitBucketMaxSize) {
+                return false;
+            }
+            else {
+                tokenBucketPermits.removeIf(ts -> ts + rateLimitDuration < System.nanoTime());
```

Comment:
how do we determine a sane value for rateLimitDuration?
Is this the sliding window algorithm for rate limiting?